File: PracticaParcialTS/Ejercicio2.ts
//Sistema para calcular los gastos en una empresa
class Gasto {
    protected descripcion: string;
    protected monto: number;

    constructor(descripcion: string, monto: number) {
        this.descripcion = descripcion;
        this.monto = monto;
    }

    mostrar(): string {
        return `${this.descripcion}: $${this.monto}`;
    }
}

class GastoCompra extends Gasto {
    constructor(producto: string, monto: number) {
        super(`Gasto por compra de ${producto}`, monto);
    }
}

class GastoServicio extends Gasto {
    constructor(compania: string, monto: number) {
        super(`Gasto por servicio de la compañía '${compania}'`, monto);
    }
}

class GastoExtraordinario extends Gasto {
    constructor(descripcion: string, monto: number) {
        super(`Gasto extraordinario '${descripcion}'`, monto);
    }
}

// Ejemplo de uso
const gastos: Gasto[] = [
    new GastoCompra("productos de limpieza", 3000),
    new GastoServicio("ServiLimp", 2500),
    new GastoExtraordinario("Medialunas para la reunión", 100)
];

gastos.forEach(gasto => {
    console.log(gasto.mostrar());
});

File: PracticaDeEjerciciosParcial/Ejercicio2.py
class Gasto:
    def __init__(self, descripcion: str, monto: float):
        self.descripcion = descripcion
        self.monto = monto

    def mostrar(self) -> str:
        return f"{self.descripcion}: ${self.monto}"

class GastoCompra(Gasto):
    def __init__(self, producto: str, monto: float):
        super().__init__(f"Gasto por compra de {producto}", monto)

class GastoServicio(Gasto):
    def __init__(self, compania: str, monto: float):
        super().__init__(f"Gasto por servicio de la compañía '{compania}'", monto)

class GastoExtraordinario(Gasto):
    def __init__(self, descripcion: str, monto: float):
        super().__init__(f"Gasto extraordinario '{descripcion}'", monto)

# Ejemplo de uso
gastos = [
    GastoCompra("productos de limpieza", 3000),
    GastoServicio("ServiLimp", 2500),
    GastoExtraordinario("Medialunas para la reunión", 100)
]

for gasto in gastos:
    print(gasto.mostrar())
